### DISCUSSION TITLE
Is there support for existing tables?

### DISCUSSION THREAD

**Discussion by kunalfleethawks** [sensedeep/dynamodb-onetable#51] [Q&A] [answered]
I really liked this library  and its very easy to use for Single Table design. But I have existing tables from prehistoric age. For new features I am using this library. But there are use cases where I want to read data from existing tables. Does this library support this use case or is there better alternative approach?

**Comment by mobsense** [comment]
OneTable does not store the schema in the table, so if you can describe the legacy data in terms of a OneTable schema, it should work fine.

Can you describe the schema for the table you want to read from?

**Reply by kunalfleethawks** [reply]
Here is the snippet
![image](https://user-images.githubusercontent.com/60139157/122332753-95a8c400-cf54-11eb-8d8c-634a5c2ee998.png)

**Comment by mobsense** [comment, marked as answer]
Without testing, here is a starting point:

```
export default {
    indexes: {
        primary: {
            hash: 'carrierID',
            sort: 'eventsSK',
        }
    },
    models: {
        Item: {
            pk:             { type: String },
            sk:             { type: String },

            eventId:        { type: String },
            ...
        }
    }
```

Th indexes defines your PK/SK to be carrierID + eventsSK. This is just a pure literal definition of the index attribute names.

Then, I've called your items as Item. You would define all your attributes in the Item mode.

Then to read, you can either use:

Item.get({carrierID: value, eventsSK: value})
Item.find({carrierID: value})

**Reply by kunalfleethawks** [reply]
Wonderful. Thanks I will try.

**Reply by kunalfleethawks** [reply]
I see whenever we create ITEMS using DynamoDB-onetable lib there is attribute added _type. But In case of existing items _type would  not be available. Will it work in that case?

**Reply by mobsense** [reply]
Without a _type, you cannot use the Model.* APIs or the Table high level APIs.  Fortunately, OneTable has low-level APIs in Table for this purpose. They are the Table.*Item.

With Model APIs, the _type is implicit. .eg.

    let User = table.getModel('User')
    User.find() 

This will apply the type defined in the schema.

With Table APIs, you provide the type as the first parameter.

    table.find('User', ...

With the Table low-level APIs, you provide attributes rather than properties. i.e. you provide PK, SK.

    table.queryItems({pk: pkvalue, sk: skvalue})

When using Table.getItem, queryItems, deleteItems, scanItems, updateItems, you do NOT provide a "type" nor is one used. This gives full access to legacy DynamoDB data created before using OneTable.

I hope this helps clarify.

**Comment by mobsense** [comment]
A "type" field is only required to read items by type. i.e. Using a schema model or using the Table level APIs that require a model.

You can still use *Items APIs on Table without having a "type".

Of course, you could update/upgrade those items and add a type field that matches a model and then you could read/write them using a model.

**Reply by kunalfleethawks** [reply]
I struggled a bit. Can you provide me example  of *Items APIs on Table without having a "type". ?

```
  this.usersTable = new Table({
            name: TableNameConstants.USERS_TABLE,
            client: this.dynamoDbClient,
            logger: this.logger(),
            delimiter: '#',
            schema: {
                indexes: {
                    primary: { hash: 'carrierID', sort: 'userSK' },
                    'userName-index': { hash: 'userName', follow: true },
                },
                models: {
                    Users: this.modelService.getUsersModel(),
                },
            },
        });
```

**Comment by mobsense** [comment]
```
const table = new Table({
    name: 'DebugTest',
    client: Client,
    schema: DebugSchema,
})
let item = await table.getItem({
    pk: `carrierID#${some-carrier-id}`,
    sk: 'userSK#',
}, {parse: true, hidden: true})
```

This will use the PK/SK as supplied. It will parse the response and marshal into JavaScript properties. The `hidden` option says to return the PK and SK properties which are normally hidden.

**Reply by mobsense** [reply]
I've also checked into `./test/debug.ts` a (temporary) unit test that show this.

**Reply by FH-sandeepkaur1** [reply]
Can I use already defined index (userName-index) instead of partition key and sort key with table.getItem?

**Reply by mobsense** [reply]
Yes. you should be able to:

1. Define the index in the schema indexes section
2. Then use the index in the params

E.g.
```
table.queryItems({pk: userName}, {index: 'userName-index', log: true})
```

The log: true will emit the actual command to the console so you can see what is generated. Post that here if it does not do what you want.

**Reply by FH-sandeepkaur1** [reply]
Here is the snippet of table definition and index definition with index name - userName-index

```
const usersTable = new Table({
            name: USERS_TABLE,
            client: this.dynamoDbClient,
            logger: this.logger(),
            delimiter: '#',
            schema: {
                indexes: {
                    primary: { hash: 'carrierID', sort: 'userSK' },
                    'userName-index': { hash: 'userName', follow: true },
                },
                models: {
                    Users: {
                        carrierID: { type: String },
                        userSK: { type: String, value: 'USR#${id}' },
                        currentStatus: {
                            type: String,
                            enum: ['Active', 'Inactive'],
                            default: 'Active',
                        },
                        email: { type: String },
                        firstName: { type: String },
                        lastName: { type: String },
                        phone: { type: String },
                        roleID: { type: String },
                        userName: { type: String },
                        userType: { type: String },
                        id: { type: String, default: ksuid.randomSync().string },
                    }
                },
            },
        });
```

When I try to invoke following query as suggested by you I'm getting error empty hash key
```
const result = await usersTable.queryItems(
                { pk: 'sandeepk'},
                {index: 'userName-index'}
            );
```

Error Output:
```
Error: dynamo: Empty hash key
    at Expression.prepare (C:\FH\fh-account-service\node_modules\dynamodb-onetable\dist\cjs\Expression.js:137:19)
    at new Expression (C:\FH\fh-account-service\node_modules\dynamodb-onetable\dist\cjs\Expression.js:19:18)
    at Model.<anonymous> (C:\FH\fh-account-service\node_modules\dynamodb-onetable\dist\cjs\Model.js:604:30)
    at Generator.next (<anonymous>)
    at C:\FH\fh-account-service\node_modules\dynamodb-onetable\dist\cjs\Model.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (C:\FH\fh-account-service\node_modules\dynamodb-onetable\dist\cjs\Model.js:4:12)
    at Model.queryItems (C:\FH\fh-account-service\node_modules\dynamodb-onetable\dist\cjs\Model.js:601:16)
    at Table.<anonymous> (C:\FH\fh-account-service\node_modules\dynamodb-onetable\dist\cjs\Table.js:454:39)
    at Generator.next (<anonymous>)
```

**Reply by mobsense** [reply]
In the schema, you have

```
                indexes: {
                    primary: { hash: 'carrierID', sort: 'userSK' },
                    'userName-index': { hash: 'userName', follow: true },
                },
```

Which means your PK is "userName".  Confirm that this is the actual attribute name of the primary key.  Remove the "follow: true" that is for following a secondary index automatically which is not what you want.

Then, you need to use "userName" in your query. Not "pk".

```
const result = await usersTable.queryItems({ userName: 'sandeepk'}, {index: 'userName-index'})
```

**Reply by kunalfleethawks** [reply]
Thanks for reply. We even tried this. no matter what we do it says empty hash key.

Our implementation:

```
this.usersTable = new Table({
            name: TableNameConstants.USERS_TABLE,
            client: this.dynamoDbClient,
            logger: (type, message, context) => {

                console.log(`${new Date().toLocaleString()}: ${type}: ${message}`);
                console.log(JSON.stringify(context, null, 4) + '\n');
            },
            delimiter: '#',
            schema: {
                indexes: {
                    primary: { hash: 'carrierID', sort: 'userSK' },
                    'userName-index': { hash: 'userName' },
                },
                models: {
                    Users: this.modelService.getUsersModel(),
                },
            },
        });

        const result = await this.usersTable.queryItems({ userName: 'sandeepk' }, { index: 'userName-index' })
        console.log(result);

```

**Reply by mobsense** [reply]
Okay, I'll test out your snippet and see.

**Reply by mobsense** [reply]
Thanks for that code sample, we've been able to reproduce the issue.

Mocking up your example, there are a couple of blocking issues. We'll pursue this and post back here in a few days with a solution.

**Reply by mobsense** [reply]
Please try with 1.5.0.

We tried this based on your example and it seemed to work

```javascript
const table = new Table({
    name: "MockTable",
    client: Client,
    logger: true,
    schema: {
        indexes: {
            primary: { hash: 'userName' },
        },
        models: {}
    }
})

await table.createTable()
let result = await table.putItem({
        userName: 'ralph', 
        email: 'ralph@example.com',
    }, {parse: true})

result = await table.scanItems({}, {parse: true})
result = await table.queryItems({userName: 'ralph'}, {parse: true})

await table.deleteTable('DeleteTableForever')
```

**Reply by kunalfleethawks** [reply]
Thanks for the reply. I tested it but it only works on primary index not on secondary index.  It does not work for GSI's 

Here is the reproducible snipped copied from your repo:

```
/*
    Demonstrate simple CRUD with OneTable
 */
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { Table } from "dynamodb-onetable";
//  For AWS V3
import Dynamo from "dynamodb-onetable/Dynamo";

const client = new Dynamo({
  client: new DynamoDBClient({
    region: "local",
    endpoint: "http://localhost:8000",
  }),
});
/*
    Single-table schema and setup. This is used for general access and by `createTable`
 */
const table = new Table({
  name: "TestCrud",
  client: client,
  uuid: "ulid",
  delimiter: "#",
  logger: true,
  schema: {
    indexes: {
      primary: { hash: "pk", sort: "sk" },
      emailIndex: { hash: "email" },
    },
    models: {
      User: {
        pk: { type: String, value: "user#${email}" },
        sk: { type: String, value: "user#${email}" },
        id: { type: String, uuid: true },
        email: { type: String, required: true },
        name: { type: String },
        status: { type: String, default: "active" },
        zip: { type: String },
      },
    },
  },
});
//  Create a model to manage User entities
const User = table.getModel("User");
async function main() {
  if (!(await table.exists())) {
    await table.createTable();
  }

  let user = await User.create(
    {
      email: "peter@example.com.com",
      name: "Peter Smith",
    },
    { log: true }
  ); //  Emit console trace for the command and result
  console.log("CREATED user", user);


  let users = await table.queryItems(
    { pk: "user#peter@example.com.com" },
    { parse: true }
  );
  console.log("Query on Primary Index", users);


  users = await table.queryItems(
    { email: "peter@example.com.com" },
    { index: "emailIndex", parse: true }
  );
  console.log("Query on GSI Index", users);
}

main();

```
